flask_common/documents.py

```diff
@@ -198,82 +221,108 @@ def setattr_unchanged(obj, key, val):
+        elif isinstance(field, ReferenceField):
+            ids = { getattr(obj, field_name).pk for obj in objs if getattr(obj, field_name, None) and getattr(getattr(obj, field_name), '_lazy', False) }
 
-        if refs:
-            if not document_class in cache_map:
-                rel_obj_map = cache_map[document_class] = {}
-            else:
-                rel_obj_map = cache_map[document_class]
+        # remove ids of objects that are already in the cache map
+        if document_class in cache_map:
+            ids -= set(cache_map[document_class])
 
-            if not document_class in id_set_map:
-                id_set = id_set_map[document_class] = set()
-            else:
-                id_set = id_set_map[document_class]
+        # no point setting up the data structures for fields where there's nothing to fetch
+        if not ids:
+            continue
 
-            # Never fetch already fetched objects
-            id_set.update(set(refs) - set(rel_obj_map.keys()))
+        # set up a cache map for the newly seen document class
+        if document_class not in cache_map:
+            cache_map[document_class] = {}
+            partial_cache_map[document_class] = {}
 
-    # Fetch objects
-    for document_class, id_set in id_set_map.iteritems():
-        rel_obj_map = cache_map[document_class]
+        # set up a fetch map for this document class
+        if document_class in fetch_map:
+            fetch_map[document_class]['ids'] |= ids
 
-        if id_set:
-            id_set = list(id_set)
-            if len(id_set) == 1:
-                rel_obj_map.update({ id_set[0]: document_class.objects.filter(pk=id_set[0]).first() })
-            else:
-                rel_obj_map.update(
-                    document_class.objects.in_bulk(id_set)
-                )
+            # make sure we don't allow partial fetching if the same document class
+            # has conflicting fields_to_fetch (e.g. { user: ["id"], created_by: True })
+            # TODO this could be improved to fetch a union of all requested fields
+            if fields_to_fetch !=  fetch_map[document_class]['fields_to_fetch']:
+                raise RuntimeError('Cannot specify different fields_to_fetch for the same document class %s' % document_class)
+        else:
+            fetch_map[document_class] = {
+                'ids': ids,
+                'fields_to_fetch': fields_to_fetch
+            }
+
+    # Fetch objects and cache them
+    for document_class, fetch_opts in fetch_map.iteritems():
+        qs = document_class.objects.filter(pk__in=fetch_opts['ids']).clear_initial_query()
+
+        # only fetch the requested fields
+        if fetch_opts['fields_to_fetch']:
+            qs = qs.only(*fetch_opts['fields_to_fetch'])
+
+        # since ids are unique, we can set a limit equal to the length
+        # of the ids set (helps with performance)
+        qs = qs.limit(len(fetch_opts['ids']))
```

Comment:
I intuitively thought it could help, but that's not true for a PK lookup. Will remove.